### PR TITLE
Partially implement type ascription

### PR DESCRIPTION
**Pull request by petrochenkov**
This PR is a rebase of the original PR by @eddyb https://github.com/rust-lang/rust/pull/21836 with some unrebasable parts manually reapplied, feature gate added + type equality restriction added as described below.

This implementation is partial because the type equality restriction is applied to all type ascription expressions and not only those in lvalue contexts. Thus, all difficulties with detection of these contexts and translation of coercions having effect in runtime are avoided.
So, you can't write things with coercions like `let slice = &[1, 2, 3]: &[u8];`. It obviously makes type ascription less useful than it should be, but it's still much more useful than not having type ascription at all.
In particular, things like `let v = something.iter().collect(): Vec<_>;` and `let u = t.into(): U;` work as expected and I'm pretty happy with these improvements alone.

Part of https://github.com/rust-lang/rust/issues/23416

### CODE REVIEW

**Comment by petrochenkov** [comment]
Hm, @rust-highfive failed to assign a reviewer.

r? @eddyb

**Comment by eddyb** [comment]
I wrote the original version of this code, so I'd have have someone else look over it.
r? @nikomatsakis

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -2671,6 +2671,14 @@ fn check_lit<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
     }
 }
 
+fn check_expr_eq_type<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
+                                     expr: &'tcx hir::Expr,
```
Comment (oli-obk) [inline src/librustc_typeck/check/mod.rs]: wrong indentation

**Comment by petrochenkov** [comment]
Rebased.

**Comment by petrochenkov** [comment]
Rebased.

**Comment by nikomatsakis** [comment]
Sorry for the delay! I'll get to this soon.

File: src/librustc_mir/hair/cx/expr.rs [hunk 2]
```diff
@@ -320,6 +320,8 @@ impl<'tcx> Mirror<'tcx> for &'tcx hir::Expr {
                                   name: Field::new(index.node as usize) },
             hir::ExprCast(ref source, _) =>
                 ExprKind::Cast { source: source.to_ref() },
+            hir::ExprType(ref source, _) =>
+                return source.make_mirror(cx),
```
Comment (nikomatsakis) [inline src/librustc_mir/hair/cx/expr.rs]: This doesn't seem right. I think we probably want to convert this to a `ExprKind::Cast`.

Comment (nikomatsakis) [inline src/librustc_mir/hair/cx/expr.rs]: /me reconsiders. Actually, let me read a bit more and get back to this point!

**Comment by nikomatsakis** [comment]
@petrochenkov I didn't notice at first that you skipped over the coercion aspect. Makes sense for a first pass, this certainly simplifies things. The code looks great. The only thing I would want to add are some parsing tests that test the relative precedence of `:` with other operators. Can you do that?

**Comment by nikomatsakis** [comment]
Some examples tests for precedence:

- `&[1, 2, 3]: &[u8; 3]` => `(&[1,2,3]): &[u8: 3]`
- `x:u8+y` => `(22:u8)+42`
- `x+y:u8` => `x+(y:u8)`

The only tricky part is how to write such at est. I envisioned testing this by setting up some scenarios that will fail to type-check if the precedence is not correct? I'm not sure if there is a better way.

**Comment by petrochenkov** [comment]
>Sorry for the delay! I'll get to this soon.

Thanks! In fact my free time is coming to an end rapidly, so I have to wrap up my work on rustc ASAP (at least for now).
I'll add tests for operator precedence.

**Comment by nikomatsakis** [comment]
@petrochenkov 

> Thanks! In fact my free time is coming to an end rapidly, so I have to wrap up my work on rustc ASAP (at least for now).

:( Let me know if there is work you think you will not have time for! I'd be happy to land this PR (or privacy) and then do the followup work myself.

**Comment by petrochenkov** [comment]
Updated with the operator precedence tests.

**Comment by nikomatsakis** [comment]
@bors r+

Another nice PR. Thanks @petrochenkov!

**Comment by petrochenkov** [comment]
This should probably be marked with `relnotes`

**Comment by brson** [comment]
[This ended up causing breakage in macros with multiple patterns](https://github.com/rust-lang/rust/issues/30531).

**Pull request merged**